File: 백준/Platinum/1786. 찾기/찾기.py
import sys
input = sys.stdin.readline

def kmp(S, W):
    T=[0 for _ in range(len(W)+1)]
    pos,cnd=1,0
    T[0]=-1
    while pos<len(W):
        if W[pos]==W[cnd]:
            T[pos]=T[cnd]
        else:
            T[pos]=cnd
            while cnd>=0 and W[pos]!=W[cnd]:
                cnd=T[cnd]
        pos+=1
        cnd+=1
    T[pos]=cnd

    j=k=0
    P=[]
    while j<len(S):
        if W[k]==S[j]:
            j+=1
            k+=1
            if k==len(W):
                P.append(j-k)
                k=T[k]
        else:
            k=T[k]
            if k<0:
                j+=1
                k+=1
    return P

if __name__=='__main__':
    T=input().rstrip()
    P=input().rstrip()
    res=kmp(T,P)
    res =[v+1 for v in res]
    print(len(res))
    print(*res)

File: 백준/Platinum/1786. 찾기/찾기.cc
#include <iostream>
#include <vector>
#include <string>
using namespace std;
vector<int> kmp(string S, string W) {
	vector<int> T(W.size()+1, 0);
	int pos = 1, cnd = 0;
	T[0] = -1;
	while (pos < W.size()) {
		if (W[pos] == W[cnd])T[pos] = T[cnd];
		else {
			T[pos] = cnd;
			while (cnd >= 0 and W[pos] != W[cnd])cnd = T[cnd];
		}
		++pos;
		++cnd;
	}
	T[pos] = cnd;

	int j = 0, k = 0;
	vector<int> P;
	while (j < S.size()) {
		if (W[k] == S[j]) {
			++j;
			++k;
			if (k == W.size()) {
				P.push_back(j - k);
				k = T[k];
			}
		}
		else {
			k = T[k];
				if (k < 0) {
					++j;
					++k;
				}
		}
	}
	return P;
}
int main() {
	string T,P;
	getline(cin, T);
	getline(cin, P);
	vector<int> res;
	res = kmp(T, P);
	for (int i = 0; i < res.size(); i++)
		++res[i];
	cout << res.size() << endl;
	for (int i = 0; i < res.size(); i++) {
		cout << res[i];
		if (i < res.size() - 1)
			cout << " ";
	}
	cout << endl;
	return 0;
}